### PR TITLE
Threading and timing decorators

### PR DESCRIPTION
**Pull request by lovesh**
Signed-off-by: lovesh harchandani <lovesh.bond@gmail.com>

### CODE REVIEW

File: nodejs/ui/views/index.ejs [hunk 1]
```diff
@@ -250,7 +250,7 @@
             function validateProof(div, proof) {
                 div.innerHTML = '<i class="fa fa-spinner fa-spin" style="font-size:100px;"></i>'
                 let xhttp = new XMLHttpRequest();
-                xhttp.open('POST', '/api/proofs/validate', true);
+                xhttp.open('POST', '/api/proofs/check_for_attrs', true);
```
Comment (dbluhm) [inline nodejs/ui/views/index.ejs]: I don't think you intended to change this file in the nodejs agent

File: test_suite/tests/conftest.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+from indy import did
+
+from modules.connection import Connection
+from modules.basicmessage import BasicMessage
+from modules.trustping import TrustPing
+from agent import Agent
+
+
+@pytest.mark.asyncio
+async def ref_agent(event_loop):
```
Comment (dbluhm) [inline test_suite/tests/conftest.py]: Having an agent that we can start and use inside the test suite is convenient for testing but I don't think the dependency of the test suite on the reference agent is appropriate, as discussed in PR #104.

Comment (lovesh) [reply test_suite/tests/conftest.py]: I mistakenly committed this. Thanks for pointing out.

File: nodejs/ui/views/index.ejs (hunk 1 above)
Comment (lovesh) [reply nodejs/ui/views/index.ejs]: You are correct. Thanks. Fixed

**Comment by lovesh** [comment]
@dbluhm I unintentionally committed some changes which had the dependencies. I removed them now

File: python/modules/__init__.py [hunk 3]
```diff
@@ -11,3 +12,22 @@ def build_problem_report_for_connections(family, problem_code, problem_str):
...
+        """
+        try:
+            msg.validate_common_blocks()
+            return 1
+        except ValidationException as e:
+            their_did = msg.context.get('from_did')
+            if their_did:
+                err_msg = self.build_problem_report_for_connections(family,
+                                                                    e.error_code, str(e.exception))
+                await self.agent.send_message_to_agent(their_did, err_msg)
```
Comment (dbluhm) [inline python/modules/__init__.py]:
```
indy-agent/python/modules/__init__.py", line 28
    await self.agent.send_message_to_agent(their_did, err_msg)
    ^
SyntaxError: 'await' outside async function
```

File: python_agent_utils/messages/fields.py [hunk 4]
```diff
@@ -0,0 +1,497 @@
+import ipaddress
+import json
+from abc import ABCMeta, abstractmethod
+from typing import Optional, Iterable
+
+import base58
+import dateutil.parser
```
Comment (dbluhm) [inline python_agent_utils/messages/fields.py]: Is this a newly introduced dependency that hasn't made it into `requirements.txt`?
```
tests/__init__.py:12: in <module>
    from python_agent_utils.messages.message import Message
../python_agent_utils/messages/message.py:11: in <module>
    from python_agent_utils.messages.fields import NonNegativeNumberField, MapField, DIDField, ISODatetimeStringField
../python_agent_utils/messages/fields.py:7: in <module>
    import dateutil.parser
E   ModuleNotFoundError: No module named 'dateutil'
```

File: python/modules/__init__.py (hunk 3 above)
Comment (lovesh) [reply python/modules/__init__.py]: Fixed this. But my tests fail now. Let me check. I did some final reorg after my tests were working and didn't check after the reorg

File: python_agent_utils/messages/fields.py (hunk 4 above)
Comment (lovesh) [reply python_agent_utils/messages/fields.py]: Added

**Comment by dbluhm** [comment]
When the test suite sends a connection response, I'm seeing the following error on the reference agent side:
```
Validation error while parsing message: {} {'@type': 'did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/connections/1.0/response', '@id': '2283d986-074f-445d-b222-03c7a18a086d', '~thread': {'thid': 'd699a559-f7c7-4b2b-b0ea-64876a8817db'}, 'connection~sig': {'@type': 'did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/signature/1.0/ed25519Sha512_single', 'signer': '72EcXPvTsy9m8ENWbHKm4dT8Att4MQ71stwZPpSmMY89', 'sig_data': 'AAAAAFzMhBR7ImRpZCI6ICJQd2h2TVRhR25WQ1laY0M1OERGNkQyIiwgImRpZF9kb2MiOiB7IkBjb250ZXh0IjogImh0dHBzOi8vdzNpZC5vcmcvZGlkL3YxIiwgImlkIjogIlB3aHZNVGFHblZDWVpjQzU4REY2RDIiLCAicHVibGljS2V5IjogW3siaWQiOiAiUHdodk1UYUduVkNZWmNDNThERjZEMiNrZXlzLTEiLCAidHlwZSI6ICJFZDI1NTE5VmVyaWZpY2F0aW9uS2V5MjAxOCIsICJjb250cm9sbGVyIjogIlB3aHZNVGFHblZDWVpjQzU4REY2RDIiLCAicHVibGljS2V5QmFzZTU4IjogIkRXSEUybkFMcmY0YVhXUWFBMjY0U2pIdERxVjFnWTZzR1d2c1BlbU1GU1dFIn1dLCAic2VydmljZSI6IFt7ImlkIjogIlB3aHZNVGFHblZDWVpjQzU4REY2RDI7aW5keSIsICJ0eXBlIjogIkluZHlBZ2VudCIsICJyZWNpcGllbnRLZXlzIjogWyJEV0hFMm5BTHJmNGFYV1FhQTI2NFNqSHREcVYxZ1k2c0dXdnNQZW1NRlNXRSJdLCAic2VydmljZUVuZHBvaW50IjogImh0dHA6Ly9sb2NhbGhvc3Q6MzAwMC9pbmR5In1dfX0=', 'signature': 'OOWYDT1OBVJ7pnoPxTKQrOENijmLyP4k8RcNLp9Ti_rPiPcq78pIu5QOyOkuG9UNqAuV2_WpCyV0jkIm7CeMCQ=='}}
```

**Comment by lovesh** [comment]
@dbluhm The newer validation rules made some older tests fail. I have pushed some changes. Will check remaining tests. We should probably think of less manual way to run these tests as we have to do it for CI. Maybe a script that interacts with the HTTP API and pytest.

**Comment by dbluhm** [comment]
> We should probably think of less manual way to run these tests as we have to do it for CI.

Yeah, manual testing is painful. We've had a few discussions on how to automate testing; @dhh1128 or @TelegramSam may need to correct me but if I recall, the consensus was to establish a connection between the test suite and agent under test and designate that connection as the "orchestrator," essentially allowing the test suite to pretend to be the identity owner. We have a prototype of how we think an identity owner might interact with an agent remotely in the reference agent as the "admin messages."

**Comment by dbluhm** [comment]
Are any of the tests still failing for you? `test_message_with_bad_timing_data` is failing for me.

**Comment by dbluhm** [comment]
@lovesh I submitted a PR to your fork with a fix for the failing tests.

**Pull request merged**